lib/mix/lib/mix/scm/git.ex

```diff
@@ -126,17 +126,17 @@ defmodule Mix.SCM.Git do
     update_origin(opts[:git])
 
     # Fetch external data
-    branch_or_tag = opts[:branch] || opts[:tag]
+    rev = get_lock_rev(opts[:lock], opts) || get_opts_rev(opts)
 
     ["--git-dir=.git", "fetch", "--force", "--quiet"]
     |> Kernel.++(progress_switch(git_version()))
     |> Kernel.++(tags_switch(opts[:tag]))
     |> Kernel.++(depth_switch(opts[:depth]))
-    |> Kernel.++(if branch_or_tag, do: ["origin", branch_or_tag], else: [])
+    |> Kernel.++(refspec_switch(opts, rev))
```

Comment:
I would merge refspec_switch and depth_switch logic into a single function. There is too much dependency between these two and unifying it in a single place would probably be cleaner.

Reply:
Other than moving the call to `depth_switch` into `refspec_switch`, I'm not seeing a clean path forward here:

```ex
  defp refspec_switch(_opts, nil), do: []

  defp refspec_switch(opts, rev) do
    depth_switch(opts[:depth]) ++
      case Keyword.take(opts, [:depth, :branch, :tag]) do
        [_ | _] -> ["origin", rev]
        _ -> []
      end
  end

  defp depth_switch(nil), do: []

  defp depth_switch(n) when is_integer(n) and n > 0 do
    check_depth_support(git_version())
    ["--depth=#{n}"]
  end
```

We only want to call `check_depth_support` with a valid `:depth`, and the other logic is about when to include the repository + refspec (e.g. origin main). Do you see a better way to organize this?

Reply:
I was thinking this:

```
  defp fetch_args(opts, nil), do: []
  
  defp fetch_args(opts, rev) do
    cond do
      depth = opts[:depth] ->
        check_depth_support(git_version())
        ["--depth=#{n}", "origin", rev]

      # We can only fetch full refs, which :ref may not be one of
      opts[:ref] == rev ->
        []

      true ->
        ["origin", rev]
    end
  end
```

Reply:
Hmm this misses the case of depth with no rev (tip of default branch, when neither of :ref, :branch, or :tag are specified)?

(I think a test case would go red)

Reply:
Isn’t that rev equals to nil, handled in the first clause?

Reply:
Ah, my code above also falls into the same mistake. The current commit handles that case with the two independent functions.

Reply:
So which arguments would reproduce the failure?

Reply:
I ran the tests locally for code in https://github.com/elixir-lang/elixir/pull/13713#discussion_r1670945353 and https://github.com/elixir-lang/elixir/pull/13713#discussion_r1671065023 (with a small fix, `s/{n}/{depth}/`).

Both fail 5 tests (5 out of 10 in the `describe "Git depth option"`): all the tests where the mix spec doesn't include one of `:ref, :branch, :tag` (in the test file they are the first and the last four tests in the describe block).

The failure is the same, the produced repos are not shallow:

     Expected truthy, got false
     code: assert File.exists?(repo_path <> "/.git/shallow")


That confirms what I said in https://github.com/elixir-lang/elixir/pull/13713#discussion_r1671104850 last night (I was on mobile and "ran the tests" on my head).

Reply:
Thanks, I got it. We need to run depth even if there is no rev. :)